scripts/utilities/param_space_exploration_utilities.py

```diff
@@ -1,85 +1,246 @@
+    else:
+        return
+    
+    
+
+    # save the paths to the run scipts for running them later
+    run_scripts = []
+
+    # create full parameter set for each run
+    # as well as the directory suffix for each run sub directory
+    for current_variable_params in all_parameter_combos(variable_params):
+        params_to_add, directory_suffix = current_variable_params
+        parameters = {**fixed_params, **params_to_add}
+        this_run = rescal_utilities.Design_a_run()
+        #this_run.set_header(run_header)
+        this_run.set_name(run_name)
+        # create a directory for each run inside the experiment directory
+        run_directory = os.path.join(experiment_directory_name, directory_suffix)
+        if not os.path.isdir(run_directory):
+            os.mkdir(run_directory)
+
+        this_run.set_directory(run_directory)
+        this_run.set_parameters(parameters)
+        this_run.write()
+
+        # store all the run scripts for future use
+        run_scripts.append(os.path.join(run_directory, run_name + '.run'))
+
+    return run_scripts
+
+
+#### NOTE: set up for power-lab machine usage, meaning on a single PC, not a cluster
+def run_rescals(run_scripts):
+    '''Takes path to a set of run_scripts that should be in the proper locations
+       to run rescal and runs each of them. Runs should be asynchronous.'''
+
+    # if any spaces in path names, turns ' ' into '\ ' so the shell can understand them
+    modded_scripts = []
+    for run_script in run_scripts:
+        modded_scripts.append(run_script.replace(' ', '\\ '))
+
+    processes = []
+    # move to the directory of each run_script and run it
+    for modded_script in modded_scripts:
+        processes.append(subprocess.Popen('cd ' + os.path.dirname(modded_script) + \
+                                          ' && ' +  modded_script, shell=True))
+    for p in processes:
+        p.wait()
+
+
+# given a top directory top_dir
+# all of its subdirectories will be checked for files
+# that match some the path_glob
+# so any file of name top_dir/"any subdirectory of top_dir"/path_glob
+# will match unless that file name matches some glob_exclude
+# examples:
+# for cellspace files path_glob='*.csp*' exclude_globs=['DUN.csp']
+# for ALTI files      path_glob='ALTI*'  exclude_globs=[]
+# TODO, 2 copies of this function, the other in param_space_exploration_utilities.py
+def get_files_to_process(top_dir, path_glob, exclude_globs):
+    
+    # get all the directories in top_directory (from stack overflow 973473)
+    # make them absolute paths
+    output_dirs = sorted([f.path for f in os.scandir(top_dir) if f.is_dir()])
+    for i in range(len(output_dirs)):
+        output_dirs[i] = os.path.abspath(output_dirs[i])
+
+    # make 2D array or absolute paths to all the .csp* files
+    paths = []
+    for output_dir in output_dirs:
+        # get all possible matches
+        local_paths = glob.glob(output_dir + '/' + path_glob)
```

Comment:
os.path.join

Reply:
fixed

Reply:
I had to reindent some stuff here and there so I think there's a decent amount of code that's showing up as changed that wasn't really.